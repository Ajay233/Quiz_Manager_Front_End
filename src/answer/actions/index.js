import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { sessionExpired } from '../../utils/session'
import history from '../../history'

// create
export const addAnswer = (data, jwt) => {
  return (dispatch) => {
    return post("answer/create", [data], jwt).then((response) => {
      dispatch({
        type: "ADD_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer created", "success", true));
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create answer", "error", true));
      }
    });
  }
}

// read
export const getAnswers = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("answer/findByQuestionId", param, jwt).then((response) => {
      dispatch({
        type: "SET_ANSWERS",
        payload: response.data
      });
    }).catch((error) => {
      dispatch(setNotification("Error retrieving answers", "error", true));
    });
  }
}

// update (there was a call to setCurrentAnswer but have removed this as it's not needed)
export const updateAnswer = (body, jwt) => {
  return (dispatch) => {
    return put("answer/update", [body], jwt).then((response) => {
      dispatch({
        type: "UPDATE_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer updated", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to update answer", "error", true));
      }
    });
  }
}

// delete
export const deleteAnswer = (config, jwt) => {
  return (dispatch) => {
    return del("answer/delete", config, jwt).then((response) => {
      dispatch(hideModal())
      dispatch({
        type: "DELETE_ANSWER",
        payload: config.data[0]
      })
      dispatch(setNotification("Answer deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(hideModal())
        dispatch(setNotification("Error - Unable to delete this answer", "error", true))
      }
    });
  }
}

export const setCurrentAnswer = (answer) => {
  return {
    type: "SET_CURRENT_ANSWER",
    payload: answer
  }
}
