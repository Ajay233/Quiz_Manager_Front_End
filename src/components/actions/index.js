export const setToTopButton = (show=false) => {
  return {
    type: "SHOW_TO_TOP_BUTTON",
    payload: show
  }
}

export const setUrl = (url="") => {
  return {
    type: "SET_URL",
    payload: url
  }
}

export const enableLightMode = () => {
  return {
    type: "SET_THEME",
    payload: false
  }
}

export const enableDarkMode = () => {
  return {
    type: "SET_THEME",
    payload: true
  }
}

export const setLoaderState = (show=false, message="", label="") => {
  return {
    type: "SET_LOADER_STATE",
    payload: {
      show: show,
      message: message,
      label: label
    }
  }
}
