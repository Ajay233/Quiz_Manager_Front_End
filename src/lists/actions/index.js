import { get } from '../../axiosRequests/requests'
import { sessionExpired } from '../../utils/session'

export const getCategories = (endpoint, jwt) => {
  return (dispatch) => {
    get(endpoint, jwt).then((response) => {
      dispatch({
        type: "SET_CATEGORIES",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      }
    })
  }
}
