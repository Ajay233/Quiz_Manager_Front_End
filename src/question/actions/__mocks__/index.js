export const getQuestions = jest.fn()

export const setCurrentQuestion = jest.fn()

export const deleteQuestion = jest.fn()

export const addQuestion = jest.fn()

export const updateQuestion = jest.fn()

export const clearQuestions = jest.fn()
