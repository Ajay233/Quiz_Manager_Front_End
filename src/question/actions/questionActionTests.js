import configureMockStore from 'reduc-mock-store'
import thunk from 'redux-thunk'
import {
  getQuestions,
  setCurrentQuestion,
  deleteQuestion,
  addQuestion,
  updateQuestion,
  clearQuestions
} from './index'

jest.mock("../../utils/session")
jest.mock("../../axiosRequests/axiosUtil")

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe("getQuestions", () => {
  it("should return an action that will add a list of questions the questions store", () => {
    const store = mockStore({})

    const expectedAction = {
      type: "SET_QUESTIONS",
      payload: [{ id: 1, questionNumber: 1 }, { id: 2, questionNumber: 2 }, { id: 3, questionNumber: 3 }]
    }

    return store.dispatch(getQuestions("question/findByQuizId", "data", "jwt")).then(() => {
      expect(store.getActions()[0]).toEqual(expectedAction)
    })
  })

  it("should call sessionExpired if an error occurrs and the status id 403", () => {
    const store = mockStore({})

    return store.dispatch(getQuestions("sessionExpired", "data", "Jwt")).then(() => {
      expect(sessionExpired).toHaveBeenCalledTimes(1);
    })
  })
})

describe("deleteQuestion", () => {
  it("should return an action that will delete a question from the questions store", () => {
    const expectedAction = {
      type: "DELETE_QUESTION",
      payload: { id: 1 }
    }

    const question = { id: 1 }

    expect(deleteQuestion(question)).toEqual(expectedAction)
  })
})

describe("addQuestion", () => {
  it("should return an action that will add a question to the questions store", () => {
    const expectedAction = {
      type: "ADD_QUESTION",
      payload: { id: 1 }
    }

    const question = { id: 1 }

    expect(addQuestion(question)).toEqual(expectedAction)
  })
})

describe("updateQuestion", () => {
  it("should return an action that will update a question in the questions store", () => {
    const expectedAction = {
      ype: "UPDATE_QUESTION",
      payload: { id: 1 }
    }

    expect(updateQuestion(question)).toEqual(expectedAction)
  })
})

describe("clearQuestions", () => {
  it("should return an action that will clear the questions store", () => {
    const expectedAction = {
      type: "CLEAR_QUESTIONS",
      payload: []
    }

    expect(clearQuestions()).toEqual(expectedAction)
  })
})

describe("setCurrentQuestion", () => {
  it("should return an action that will add a question the question store", () => {
    const expectedAction = {
      type: "SET_CURRENT_QUESTION",
      payload: { id: 1 }
    }

    const question = { id: 1 }

    expect(setCurrentQuestion(question)).toEqual(expectedAction)
  })
})
