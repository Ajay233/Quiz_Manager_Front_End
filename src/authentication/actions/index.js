import { post } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import history from '../../history'

export const setUser = (endpoint, loginDetails) => {

  return (dispatch) => {

    post(endpoint, loginDetails).then((response) => {
      dispatch({
        type: "SET_USER_LOGGED_IN",
        payload: {
          id: response.data.user.id,
          forename: response.data.user.forename,
          surname: response.data.user.surname,
          email: response.data.user.email,
          permission: response.data.user.permission,
          verified: response.data.user.verified,
          jwt: response.data.jwt,
          loggedIn: true,
          response: response
        }
      })
      history.push("/")
      dispatch(setNotification(`Welcome back ${response.data.user.forename}`, "loginSuccess", true))
    }).catch((error) => {
      if(error.response.data === "NOT VERIFIED"){
        const verifyMsg = "Your email has not yet been veirified.  You will need to verify your email before you can log in"
        dispatch(setNotification(verifyMsg, "warning", true))
      } else {
        dispatch({
          type: "SET_USER_LOGGED_IN",
          payload: {
            error: error.response
          }
        });
        dispatch(setNotification("The username or password you entered was incorrect", "error", true));
      }
    })
  }
}

export const logOutUser = () => {
  return {
    type: "LOG_OUT_USER",
    payload: {
      forename: "",
      surname: "",
      email: "",
      permission: "",
      verified: "",
      jwt: "",
      loggedIn: false
    }
  }
}

export const setVerficationProcess = (endpoint, verificationDetails) => {

  return (dispatch) => {

    post(endpoint, verificationDetails).then((response) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "completed",
          token: "",
          error: {}
        }
      })
    }).catch((error) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "not verified",
          token: verificationDetails.token,
          error: error.response
        }
      })
    })
  }
}

export const setRedirect = (status) => {
  return {
    type: "SET_REDIRECT",
    payload: {
      status: status
    }
  }
}

export const logOut = () => {
  return {
    type: "RESET_APP"
  }
}

// export const getUsers = (endpoint, header) => {
//   return {
//     type: "GET_USER",
//     payload: {
//       userId
//   }
// }
