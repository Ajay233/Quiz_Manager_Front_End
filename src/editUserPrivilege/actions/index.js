import { getUsingParams } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const setUserResults = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("users/findByEmail", param, jwt).then((response) => {
      dispatch({
        type: "SET_USER_RESULTS",
        payload: response.data
      })
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        const msg = error.response.data;
        dispatch(setNotification(msg, "error", true));
        console.log(error.response)
      }
    });
  }
}

export const clearUserResults = () => {
  return {
    type: "CLEAR_USER_RESULTS",
    payload: {}
  }
}
