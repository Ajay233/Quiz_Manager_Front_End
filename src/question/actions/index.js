import { reset } from 'redux-form'
import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { getAnswers } from '../../answer/actions'
import { sessionExpired } from '../../utils/session'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { setLoaderState } from '../../components/actions'
import history from '../../history'

export const addQuestion = (body, jwt) => {
  return (dispatch) => {
    dispatch(setLoaderState(true, "Saving...", "saving"))
    return post("question/create", body, jwt).then((response) => {
      dispatch(setLoaderState())
      dispatch({
        type: "ADD_QUESTION",
        payload: response.data
      })
      history.push("/editQuiz");
      dispatch(setNotification("Question created", "success", true));
    }).catch((error) => {
      dispatch(setLoaderState())
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else if(error.response.status === 500 && error.response.data.message.includes("Maximum upload size exceeded")) {
        dispatch(setNotification("Error - Image exceeds the maximum size of 1040KB", "error", true));
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create question with the details provided", "error", true))
      }
    });
  }
}

export const getQuestions = (endpoint, param, jwt, startQuiz=false) => {
  return (dispatch) => {
    return getUsingParams(endpoint, param, jwt).then((response) => {
      if(response.data !== "NO QUESTIONS"){
        dispatch({
          type: "SET_QUESTIONS",
          payload: response.data
        })
      } else {
        dispatch({
          type: "SET_QUESTIONS",
          payload: []
        })
      }
      if(startQuiz){
        const body = { questionId: response.data[0].id }
        dispatch(getAnswers(body, jwt))
        dispatch(setCurrentQuestion(response.data[0]));
      }
      console.log("success")
    }).catch((error) => {
      console.log(error);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification(error.response.data, "error", true))
      }
    });
  }
}

export const updateQuestion = (body, jwt) => {
  return (dispatch) => {
    dispatch(setLoaderState(true, "Saving...", "saving"))
    return put("question/update", body, jwt).then((response) => {
      dispatch(setLoaderState())
      dispatch(setCurrentQuestion(response.data));
      dispatch({
        type: "UPDATE_QUESTION",
        payload: response.data
      })
      dispatch(setNotification("Question updated", "success", true))
      dispatch(reset('updateQuestionForm'))
    }).catch((error) => {
      console.log(error.response);
      dispatch(setLoaderState())
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else if(error.response.status === 500 && error.response.data.message.includes("Maximum upload size exceeded")) {
        dispatch(setNotification("Error - Image exceeds the maximum size of 1040KB", "error", true));
      } else {
        dispatch(setNotification("Error - unable to update question", "error", true))
      }
    });
  }
}

export const setCurrentQuestion = (question) => {
  return {
    type: "SET_CURRENT_QUESTION",
    payload: question
  }
}

export const deleteQuestion = (config, jwt) => {
  return (dispatch) => {
    dispatch(hideModal())
    dispatch(setLoaderState(true, "Deleting..."))
    return del("question/delete", config, jwt).then((response) => {
      dispatch(setLoaderState())
      dispatch({
        type: "DELETE_QUESTION",
        payload: config.data[0]
      })
      history.push("/editQuiz")
      dispatch(setNotification("Question deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      dispatch(setLoaderState())
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - Unable to delete this question", "error", true))
      }
    });
  }
}

export const deleteImage = (config, jwt) => {
  return (dispatch) => {
    dispatch(hideModal())
    dispatch(setLoaderState(true, "Deleting..."))
    return del("question/deleteImage", config, jwt).then((response) => {
      dispatch(setLoaderState())
      dispatch(setCurrentQuestion(response.data))
      dispatch({
        type: "UPDATE_QUESTION",
        payload: response.data
      })
      dispatch(setNotification("Image has been deleted", "success", true))
    }).catch((error) => {
      console.log(error.response);
      dispatch(setLoaderState())
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - Unable to delete this image", "error", true))
      }
    })
  }
}

export const clearQuestions = () => {
  return {
    type: "CLEAR_QUESTIONS",
    payload: []
  }
}
