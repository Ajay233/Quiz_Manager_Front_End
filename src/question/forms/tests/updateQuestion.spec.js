import React from 'react'
import { Router } from 'react-router-dom'
import { Provider } from 'react-redux'
import UpdateQuestionForm from '../updateQuestion'
import { mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { render, fireEvent, cleanup } from '@testing-library/react'
import history from '../../../history'
import { setCurrentQuestion, updateQuestion } from '../../actions'

jest.mock('../../actions')

const mockStore = configureStore({})

describe("UpdateQuestionForm", () => {

  let store;

  beforeEach(() => {
    store = mockStore({
      currentQuestion: {
        id: 1,
        quizId: 1,
        questionNumber: 1,
        description: "question"
      },
      userData: {
        id: 1,
        forename: "testForename",
        surname: "testSurname",
        email: "testEmail",
        permission: "USER",
        verified: "true",
        jwt: "testJwt",
        loggedIn: true
      }
    })
  })

  it("should call action creators on submit success", () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <UpdateQuestionForm />
        </Router>
      </Provider>
    )

    wrapper.find('form').simulate('submit')
    expect(updateQuestion).toHaveBeenCalledTimes(1)
  })

  it("should call triggerModal when the delete button is clicked", () => {
    const triggerModal = jest.fn()

    const wrapper = render(
      <Provider store={store}>
        <Router history={history}>
          <UpdateQuestionForm
            triggerModal={triggerModal}
          />
        </Router>
      </Provider>
    )

    fireEvent.click(wrapper.getByTestId("delete-question-button"))
    expect(triggerModal).toHaveBeenCalledTimes(1)
  })

  it("should match the snapshot", () => {
    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <UpdateQuestionForm />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

})
