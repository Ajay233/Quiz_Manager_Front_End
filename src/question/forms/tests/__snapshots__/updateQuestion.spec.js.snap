// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateQuestionForm should match the snapshot 1`] = `
<div
  className=""
>
  <div
    className="title-large"
    id="questionViewTitle"
  >
    Edit Question 1
  </div>
  <div
    className=""
  >
    <div
      className=""
    >
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div>
          <label>
            Question Number:
          </label>
          <input
            className="inputBox"
            name="number"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            value=""
          />
        </div>
        <div>
          <label>
            Question description:
          </label>
          <textarea
            className="inputBox"
            name="description"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            value=""
          />
        </div>
        <div
          className="buttonGroup"
        >
          <button
            className="submit"
          >
            <i
              className="far fa-save"
            />
             Save Changes
          </button>
          <button
            className="delete"
            data-testid="delete-question-button"
          >
            <i
              className="fas fa-trash-alt"
            />
             Delete
          </button>
        </div>
      </form>
    </div>
    <div
      className=""
    />
  </div>
</div>
`;
